### ISSUE TITLE
pytest_assertrepr_compare() hook doesn't work with all assert syntax

### ISSUE THREAD

**Issue by geekyprats** [pytest-dev/pytest#5535]
Hi All,

There is pytest_assertrepr_compare() hook for assertion expression comparison.
However the hook will work only if assert statement has both left and right expressions along with operands.
Example
1. assert 1 == 2
2. assert i == j
3. assert 'a' not in 'assert':

My use case is - Every time an assertion failure happens, I need a hook to enter in and manage few test specific variables defined inside my conftest.
But since the hook requires an operator, it will not work for below assert statements.

1. assert i
2. assert not j
etc.

Is there a solution to this, that allows me to enter a hook on every assertion failure ?

**Comment by debugduck** [comment]
Is this still open? I'm new so I may be missing something but I tested out assert statements such as `assert i` and `assert not j` and they seemed to work as intended. I also noticed it looked like there was a rewrite already of the file where this hook lives.

**Comment by bluetech** [comment]
> Is there a solution to this, that allows me to enter a hook on every assertion failure ?

Currently there isn't.

@debugduck

> Is this still open?

Yes, the `pytest_assertrepr_compare` hook only runs for asserts using  `'<'|'>'|'=='|'>='|'<='|'<>'|'!='|'in'|'not in'|'is'|'is not'` comparisons (list taken from Python grammar `comp_op`).

The hook takes `op, left, right` arguments so it probably can't be generalized directly.

**Comment by facundobatista** [comment]
I'd love to have a hook for this.

In my [pytest plugin](https://github.com/facundobatista/logassert/), I need to have control on messages if the user does `assert myplugin.container` and it fails (currently I only have control on the messages when the user does `assert "stuff" in myplugin.container`, thanks to the `pytest_assertrepr_compare` hook.

Thanks!

**Comment by wimglenn** [comment]
I have a compelling use-case for this too in my plugin [`pytest-structlog`](https://github.com/wimglenn/pytest-structlog).  It defines a `log` fixture which you can use like this:

       assert log.has("reticulated spline", n_splines=3, level="info")

In case of `log.has` returning Falsey, it would be great to be able to check if there were any log events which were really *similar*, and print out any similar looking events.  e.g. if there was no log event "reticulated spline" found but there was one "reticulated splines", I should be able to suggest that to the user (likely typo).

Currently the failure mode is super gross:

```
    def test_fail(log):
        logger.info("bleh")
>       assert log.has("blah")
E       AssertionError: assert False
E        +  where False = <bound method StructuredLogCapture.has of <pytest_structlog.StructuredLogCapture object at 0x10c4dc790>>('blah')
E        +    where <bound method StructuredLogCapture.has of <pytest_structlog.StructuredLogCapture object at 0x10c4dc790>> = <pytest_structlog.StructuredLogCapture object at 0x10c4dc790>.has
```

**Comment by RonnyPfannschmidt** [comment]
We can consider adding a hook that interprets boolean values which are not bools/enums

That way we could have expression helpers in general

**Comment by wimglenn** [comment]
Yes that was my thought too, plug-in authors can return instances of a custom type which defines a __bool__ method.

**Comment by debugduck** [comment]
So it sounds like we want to add another hook in addition to the current one which takes some custom type that defines a `__bool__` method as input, as opposed to the current hook which restricts us to the pattern of having some operation, a left operand, and a right operand.

I'm new to the hooks,  so I apologize if this is a silly but I was wondering, since the implementation is left up to the user, is there any way other than in the comment of the hook that the return type or other rules are enforced?

**Comment by RonnyPfannschmidt** [comment]
not custom type, any object that can do boolean context will do

main problem is how do we want to handle shortcut logic operators and  negation using not or ~

**Comment by wimglenn** [comment]
The ~ is easy since the plugin author also controls `__invert__`.  But `not obj` is tricky because that will be plain old boolean regardless, unless I'm forgetting some trick.

**Comment by trewest** [comment]
Is there any update on the direction of this?
